### PR TITLE
Add force_light option to prevent regular stemcells from being used

### PR DESCRIPTION
**Pull request by jpalermo**
The current behavior of defaulting to light stemcells and returning regular stemcells only if NO light stemcells seems correct but fails when doing incremental checks.

The current behavior fails when doing a check and a single new regular stemcell has been published, but the light version has not. In that case the resource will see that there are no light stemcells in the returned list, and return the regular version.

This can lead to consuming regular stemcells when it's important that only light stemcells are used.

### CODE REVIEW

File: acceptance/check_test.go [hunk 1]
```diff
@@ -273,29 +265,6 @@ var _ = Describe("check", func() {
 				Expect(result[0]["version"]).NotTo(BeEmpty())
 			})
 		})
-
-		XContext("and only light stemcell versions are available", func() {
```
Comment (jpalermo) [inline acceptance/check_test.go:277]: Tried to add an integration test here, but realized that the situation this is fixing is an edge case, so it can't be reliably integration tested.

Removed this X'ed out test too as it has the same problem and will probably never be usable.

File: README.md [hunk 2]
```diff
@@ -25,6 +25,11 @@ would match `3262.1` and `3262.1.1`, but not `3262.2`.
 * `force_regular`: *Optional.* Default `false`. By default, the resource will always download light stemcells for IaaSes that support light stemcells.
   If `force_regular` is `true`, the resource will ignore light stemcells and always download regular stemcells.
 
+* `force_light`: *Optional.* Default `false`. By default, the resource will always download light stemcells for IaaSes
+  that support light stemcells. If `force_light` is `true`, the resource will ignore regular stemcells and always
```
Comment (xtremerui) [inline README.md:29]: what if both force_regular and force_light are set to true? From the logic I can tell force_regular take precedence here. 

Could you make it clear?

File: boshio/stemcells.go [hunk 3]
```diff
@@ -36,18 +37,14 @@ func (s Stemcells) FindStemcellByVersion(version string) (Stemcell, bool) {
...
+	} else if s.forceLight()  {
+		return s.lightStemcellsOnly()
+	} else if s.supportsLight() == false {
+		// all stemcells are Regular, no need to filter
+		return s
 	} else {
-		// The light stemcells might be published several hours after the regular versions
-		// The resource should wait until the corresponding light version is available to avoid
-		// caching a bulky regular stemcell
 		return s.lightStemcellsOnly()
```
Comment (xtremerui) [inline boshio/stemcells.go:48]: in the description of `force_light` you mention
```
Without `force_light` set it is possible for a `check` to return a regular stemcell in
  cases where the regular has been published, but the light version is not yet published.
```

this is against the code here. I wonder is `s.lightStemcellsOnly()` a wrong name since it could return regular? Or was the previous comment that got deleted above not accurate?

**Review by xtremerui — COMMENTED** [review]
Thx for the PR. Just need to sort out some descriptive words.

File: README.md (hunk 2 above)
Comment (jpalermo) [reply README.md:29]: If you set them both to true, the way the current code is structured you'll get regular stemcells. But I'd consider this undefined behavior, I can think of no use case that makes sense for setting both to true, so it would only be a mistake.

Comment (xtremerui) [reply README.md:29]: my mistake here. I meant `force_regular` will take precedence based on line https://github.com/concourse/bosh-io-stemcell-resource/blob/3d36bd8c8653ba2b5e46984d6c22f7e79d1c171a/boshio/stemcells.go#L40

so in the doc of `force_regular`, i think it is better to explicity call it out.

File: README.md [hunk 4]
```diff
@@ -25,6 +25,11 @@ would match `3262.1` and `3262.1.1`, but not `3262.2`.
 * `force_regular`: *Optional.* Default `false`. By default, the resource will always download light stemcells for IaaSes that support light stemcells.
   If `force_regular` is `true`, the resource will ignore light stemcells and always download regular stemcells.
```
Comment (xtremerui) [inline README.md:26]:
```suggestion
  If `force_regular` is `true`, the resource will ignore light stemcells (even with `force_light` set to `true`) and always download regular stemcells.
```

File: README.md (hunk 2 above)
Comment (xtremerui) [inline README.md:29]:
```suggestion
  that support light stemcells. If `force_light` is `true` and `force_regular` is not set, the resource will ignore regular stemcells and always
```

File: boshio/stemcells.go (hunk 3 above)
Comment (jpalermo) [reply boshio/stemcells.go:48]: `s.lightStemcellsOnly()` is correctly named, that function will never return a regular stemcell.

The problem previously was the `if s.supportsLight() == false {` conditional wasn't always correct.

`supportsLight` assumes that if it sees a single stemcell in the list, that light stemcells are supported. This is generally true, but falls apart when the list of stemcells is just a list of one because it's an incremental check.

The alternative way to fix this would be to determine if the stemcell `supportsLight` before the list of stemcells is filtered. But I believe currently the main filtering is happening via the bosh.io api, so this would involving pulling down the full list of stemcells from bosh.io for every check. ~~Which is probably not an idea usage of resources for the concourse workers or the bosh.io servers.~~

Actually, I was wrong, I thought it used the previous check version to filter the api call to bosh.io, but it doesn't at all, it actually does fetch all the stemcells and filters them after the fact.

I'm going to close this PR and create a new one that just fixes `supportsLight` so it is correct. This should just fix the edge case behavior, without having to change the configuration interface at all.

**Comment by jpalermo** [comment]
Thoroughly confused now...

We had a pipeline run that always gets light stemcells, and this one time it found a new heavy stemcell instead. I had assumed there was a race condition when the heavy is published before the light and the only way to fix this would be to introduce the above change.

However, I had been making the assumption that it only determines if the stemcell line supports "light" stemcells AFTER it had filtered by the previous check version. This assumption fit what we were seeing, and sort of makes sense.

Looking closer at the code now, it doesn't filter by version using the bosh.io api at all (and probably can't to be fair), but fetches all the stemcells, which means the `supportsLight` check should pretty much always work correctly.

So how did we end up with a heavy stemcell then?

Turns out bosh.io was inconsistent. 7 of the 10 workers knew about the light stemcell for a minute or two while 3 of the 10 workers hadn't yet pulled that data. Normally all the workers are updated within a couple of seconds of each other, but there were two separate updates running in quick succession, and some of the workers ended up pulling data from the second update while running the first.

So the `check` happened on one of the 7 workers that knew about the light stemcell, but then the `get` happened against one of the 3 that didn't and could only return a heavy stemcell for the requested version.

In theory we could change the `get` action to do a similar `supportsLight` check that `check` is doing. But you don't have a lot of great options at that point. You can pretty much fail, which will be a red build, or you can throw a retry loop in there.

Neither of those are great options. I'm going to spend some time today seeing if I can make bosh.io a bit more consistent when it's updating itself.
